File: BlueType.Agent.Tests/ProtocolExamplesTests.cs
using System.Reflection;
using System.Text.Json;
using BlueType.Protocol;

namespace BlueType.Agent.Tests;

public sealed class ProtocolExamplesTests
{
    public static IEnumerable<object[]> ExampleFiles()
    {
        foreach (var path in Directory.EnumerateFiles(FindExamplesDirectory(), "*.json").OrderBy(Path.GetFileName))
        {
            yield return new object[] { path };
        }
    }

    [Theory]
    [MemberData(nameof(ExampleFiles))]
    public async Task ProtocolExample_DecodesAndRoundTripsThroughFrameCodec(string path)
    {
        var manifest = await LoadManifestAsync();
        var knownTypes = manifest.Commands.Concat(manifest.Responses).ToHashSet(StringComparer.Ordinal);
        var knownErrorCodes = manifest.ErrorCodes.ToHashSet(StringComparer.Ordinal);
        var json = await File.ReadAllTextAsync(path);
        var envelope = JsonSerializer.Deserialize<Envelope>(json, JsonProtocol.SerializerOptions);

        Assert.NotNull(envelope);
        Assert.Equal(manifest.Version, envelope.V);
        Assert.False(string.IsNullOrWhiteSpace(envelope.Id));
        Assert.Contains(envelope.Type, knownTypes);
        Assert.Equal(JsonValueKind.Object, envelope.Payload.ValueKind);

        if (string.Equals(envelope.Type, Responses.Error, StringComparison.Ordinal))
        {
            Assert.True(envelope.Payload.TryGetProperty("code", out var code));
            Assert.Equal(JsonValueKind.String, code.ValueKind);
            var errorCode = code.GetString();
            Assert.NotNull(errorCode);
            Assert.Contains(errorCode, knownErrorCodes);
        }

        using var stream = new MemoryStream();
        await FrameCodec.WriteAsync(stream, envelope);
        stream.Position = 0;

        var roundTripped = await FrameCodec.ReadAsync(stream);
        Assert.NotNull(roundTripped);
        AssertEnvelopeEquivalent(envelope, roundTripped);
    }

    [Fact]
    public async Task ProtocolManifest_RequiresFixtureCoverageForEveryMessageTypeAndErrorCode()
    {
        var manifest = await LoadManifestAsync();
        var expectedTypes = manifest.Commands.Concat(manifest.Responses).ToHashSet(StringComparer.Ordinal);
        var coveredTypes = new HashSet<string>(StringComparer.Ordinal);
        var coveredErrorCodes = new HashSet<string>(StringComparer.Ordinal);

        foreach (var path in Directory.EnumerateFiles(FindExamplesDirectory(), "*.json"))
        {
            var envelope = Assert.IsType<Envelope>(JsonSerializer.Deserialize<Envelope>(
                await File.ReadAllTextAsync(path),
                JsonProtocol.SerializerOptions));

            Assert.Contains(envelope.Type, expectedTypes);
            coveredTypes.Add(envelope.Type);

            if (string.Equals(envelope.Type, Responses.Error, StringComparison.Ordinal) &&
                envelope.Payload.TryGetProperty("code", out var code) &&
                code.ValueKind == JsonValueKind.String)
            {
                coveredErrorCodes.Add(code.GetString()!);
            }
        }

        foreach (var type in expectedTypes)
        {
            Assert.Contains(type, coveredTypes);
        }

        foreach (var errorCode in manifest.ErrorCodes)
        {
            Assert.Contains(errorCode, coveredErrorCodes);
        }
    }

    [Fact]
    public async Task CSharpProtocolConstants_MatchProtocolManifest()
    {
        var manifest = await LoadManifestAsync();

        Assert.Equal(
            manifest.Commands.OrderBy(value => value, StringComparer.Ordinal),
            GetProtocolConstants(typeof(Commands)));
        Assert.Equal(
            manifest.Responses.OrderBy(value => value, StringComparer.Ordinal),
            GetProtocolConstants(typeof(Responses)));
    }

    [Fact]
    public async Task InvalidProtocolExamples_AreRejectedByV1Contract()
    {
        var manifest = await LoadManifestAsync();

        foreach (var path in Directory.EnumerateFiles(FindInvalidDirectory(), "*.json").OrderBy(Path.GetFileName))
        {
            using var document = JsonDocument.Parse(await File.ReadAllTextAsync(path));
            Assert.False(IsValidEnvelope(document.RootElement, manifest), path);
        }
    }

    private sealed record ProtocolManifest(
        int Version,
        string[] Commands,
        string[] Responses,
        string[] ErrorCodes);

    private static async Task<ProtocolManifest> LoadManifestAsync()
    {
        var manifest = JsonSerializer.Deserialize<ProtocolManifest>(
            await File.ReadAllTextAsync(Path.Combine(FindSpecDirectory(), "protocol-v1.json")),
            JsonProtocol.SerializerOptions);
        return Assert.IsType<ProtocolManifest>(manifest);
    }

    private static string[] GetProtocolConstants(Type constantsType)
    {
        return constantsType
            .GetFields(BindingFlags.Public | BindingFlags.Static)
            .Where(field => field.IsLiteral && !field.IsInitOnly && field.FieldType == typeof(string))
            .Select(field => (string)field.GetRawConstantValue()!)
            .OrderBy(value => value, StringComparer.Ordinal)
            .ToArray();
    }

    private static bool IsValidEnvelope(JsonElement root, ProtocolManifest manifest)
    {
        if (root.ValueKind != JsonValueKind.Object ||
            !root.TryGetProperty("v", out var version) ||
            version.ValueKind != JsonValueKind.Number ||
            !version.TryGetInt32(out var versionValue) ||
            versionValue != manifest.Version ||
            !HasNonEmptyString(root, "id") ||
            !HasNonEmptyString(root, "type") ||
            !root.TryGetProperty("payload", out var payload) ||
            payload.ValueKind != JsonValueKind.Object)
        {
            return false;
        }

        var type = root.GetProperty("type").GetString()!;
        if (!manifest.Commands.Concat(manifest.Responses).Contains(type, StringComparer.Ordinal))
        {
            return false;
        }

        return type switch
        {
            Commands.Hello => HasNonEmptyString(payload, "deviceId") && HasNonEmptyString(payload, "deviceName"),
            Commands.TextInsert => HasString(payload, "text"),
            Commands.KeyTap or Commands.KeyDown or Commands.KeyUp => HasNonEmptyString(payload, "key"),
            Commands.Combo => HasStringArray(payload, "keys"),
            Commands.MouseMove => HasInt(payload, "dx") && HasInt(payload, "dy"),
            Commands.MouseButton => HasNonEmptyString(payload, "button") && HasOneOf(payload, "action", "down", "up"),
            Commands.MouseClick => HasNonEmptyString(payload, "button") &&
                                   (!payload.TryGetProperty("repeat", out var repeat) || IsInt(repeat)),
            Commands.MouseScroll => OptionalInt(payload, "deltaX") && OptionalInt(payload, "deltaY"),
            Commands.ClipboardSet => HasString(payload, "text"),
            Commands.ClipboardGet or Commands.Ping or Responses.Pong => true,
            Responses.Ack => HasBoolean(payload, "ok"),
            Responses.Error => HasErrorPayload(payload, manifest),
            Responses.AuthPending => HasInt(payload, "timeoutSec") && HasString(payload, "message"),
            Responses.AuthResult => HasBoolean(payload, "ok") &&
                                    HasBoolean(payload, "persistToken") &&
                                    HasBoolean(payload, "trusted"),
            Responses.ClipboardValue => HasString(payload, "text"),
            Responses.ShortcutProfile => HasNullableString(payload, "name") && HasNullableObject(payload, "profile"),
            _ => false,
        };
    }

    private static bool HasErrorPayload(JsonElement payload, ProtocolManifest manifest)
    {
        return HasString(payload, "message") &&
               payload.TryGetProperty("code", out var code) &&
               code.ValueKind == JsonValueKind.String &&
               code.GetString() is string codeValue &&
               manifest.ErrorCodes.Contains(codeValue, StringComparer.Ordinal);
    }

    private static bool HasNonEmptyString(JsonElement element, string propertyName)
    {
        return HasString(element, propertyName) && !string.IsNullOrWhiteSpace(element.GetProperty(propertyName).GetString());
    }

    private static bool HasString(JsonElement element, string propertyName)
    {
        return element.TryGetProperty(propertyName, out var property) && property.ValueKind == JsonValueKind.String;
    }

    private static bool HasNullableString(JsonElement element, string propertyName)
    {
        return element.TryGetProperty(propertyName, out var property) &&
               property.ValueKind is JsonValueKind.String or JsonValueKind.Null;
    }

    private static bool HasNullableObject(JsonElement element, string propertyName)
    {
        return element.TryGetProperty(propertyName, out var property) &&
               property.ValueKind is JsonValueKind.Object or JsonValueKind.Null;
    }

    private static bool HasStringArray(JsonElement element, string propertyName)
    {
        return element.TryGetProperty(propertyName, out var property) &&
               property.ValueKind == JsonValueKind.Array &&
               property.EnumerateArray().All(item => item.ValueKind == JsonValueKind.String);
    }

    private static bool HasInt(JsonElement element, string propertyName)
    {
        return element.TryGetProperty(propertyName, out var property) && IsInt(property);
    }

    private static bool OptionalInt(JsonElement element, string propertyName)
    {
        return !element.TryGetProperty(propertyName, out var property) || IsInt(property);
    }

    private static bool IsInt(JsonElement property)
    {
        return property.ValueKind == JsonValueKind.Number && property.TryGetInt32(out _);
    }

    private static bool HasBoolean(JsonElement element, string propertyName)
    {
        return element.TryGetProperty(propertyName, out var property) &&
               property.ValueKind is JsonValueKind.True or JsonValueKind.False;
    }

    private static bool HasOneOf(JsonElement element, string propertyName, params string[] values)
    {
        return HasString(element, propertyName) &&
               element.GetProperty(propertyName).GetString() is string value &&
               values.Contains(value, StringComparer.Ordinal);
    }

    private static string FindExamplesDirectory()
    {
        return Path.Combine(FindSpecDirectory(), "examples");
    }

    private static string FindInvalidDirectory()
    {
        return Path.Combine(FindSpecDirectory(), "invalid");
    }

    private static string FindSpecDirectory()
    {
        var current = new DirectoryInfo(AppContext.BaseDirectory);
        while (current is not null)
        {
            var candidate = Path.Combine(current.FullName, "protocol", "spec");
            if (Directory.Exists(candidate))
            {
                return candidate;
            }

            current = current.Parent;
        }

        throw new DirectoryNotFoundException("Could not find protocol/spec from the test output directory.");
    }

    private static void AssertEnvelopeEquivalent(Envelope expected, Envelope actual)
    {
        Assert.Equal(expected.V, actual.V);
        Assert.Equal(expected.Id, actual.Id);
        Assert.Equal(expected.Type, actual.Type);
        Assert.Equal(expected.Token, actual.Token);
        Assert.True(JsonElementDeepEquals(expected.Payload, actual.Payload));
    }

    private static bool JsonElementDeepEquals(JsonElement left, JsonElement right)
    {
        using var leftDocument = JsonDocument.Parse(left.GetRawText());
        using var rightDocument = JsonDocument.Parse(right.GetRawText());
        return JsonElementDeepEqualsCore(leftDocument.RootElement, rightDocument.RootElement);
    }

    private static bool JsonElementDeepEqualsCore(JsonElement left, JsonElement right)
    {
        if (left.ValueKind != right.ValueKind)
        {
            return false;
        }

        return left.ValueKind switch
        {
            JsonValueKind.Object => ObjectsDeepEqual(left, right),
            JsonValueKind.Array => ArraysDeepEqual(left, right),
            JsonValueKind.String => left.GetString() == right.GetString(),
            JsonValueKind.Number => left.GetRawText() == right.GetRawText(),
            JsonValueKind.True or JsonValueKind.False => left.GetBoolean() == right.GetBoolean(),
            JsonValueKind.Null => true,
            _ => left.GetRawText() == right.GetRawText(),
        };
    }

    private static bool ObjectsDeepEqual(JsonElement left, JsonElement right)
    {
        var leftProperties = left.EnumerateObject().OrderBy(property => property.Name, StringComparer.Ordinal).ToArray();
        var rightProperties = right.EnumerateObject().OrderBy(property => property.Name, StringComparer.Ordinal).ToArray();
        if (leftProperties.Length != rightProperties.Length)
        {
            return false;
        }

        for (var i = 0; i < leftProperties.Length; i++)
        {
            if (!string.Equals(leftProperties[i].Name, rightProperties[i].Name, StringComparison.Ordinal))
            {
                return false;
            }

            if (!JsonElementDeepEqualsCore(leftProperties[i].Value, rightProperties[i].Value))
            {
                return false;
            }
        }

        return true;
    }

    private static bool ArraysDeepEqual(JsonElement left, JsonElement right)
    {
        var leftItems = left.EnumerateArray().ToArray();
        var rightItems = right.EnumerateArray().ToArray();
        if (leftItems.Length != rightItems.Length)
        {
            return false;
        }

        for (var i = 0; i < leftItems.Length; i++)
        {
            if (!JsonElementDeepEqualsCore(leftItems[i], rightItems[i]))
            {
                return false;
            }
        }

        return true;
    }
}

File: BlueType.Mac/Tests/BlueTypeMacCoreTests/ProtocolExamplesTests.swift
import Foundation
import XCTest
@testable import BlueTypeMacCore

final class ProtocolExamplesTests: XCTestCase {
    func testProtocolExamplesDecodeAndRoundTripThroughFrameCodec() throws {
        let manifest = try readManifest()
        let examplesDirectory = try findExamplesDirectory()
        let exampleFiles = try FileManager.default.contentsOfDirectory(
            at: examplesDirectory,
            includingPropertiesForKeys: nil
        )
            .filter { $0.pathExtension == "json" }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }

        XCTAssertFalse(exampleFiles.isEmpty)

        for file in exampleFiles {
            let payload = try Data(contentsOf: file)
            let envelope = try JSONDecoder().decode(Envelope.self, from: payload)

            XCTAssertEqual(envelope.v, manifest.version, file.lastPathComponent)
            XCTAssertFalse(envelope.id.isEmpty, file.lastPathComponent)
            XCTAssertTrue(
                (manifest.commands + manifest.responses).contains(envelope.type),
                "Unknown type \(envelope.type) in \(file.lastPathComponent)"
            )

            if envelope.type == MessageType.error {
                let code = envelope.payload["code"]?.stringValue
                XCTAssertNotNil(code, file.lastPathComponent)
                XCTAssertTrue(
                    manifest.errorCodes.contains(code ?? ""),
                    "Unknown error code \(code ?? "nil") in \(file.lastPathComponent)"
                )
            }

            let frame = try FrameCodec.encode(envelope)
            let decoded = try FrameCodec.decodePayload(frame.dropFirst(4))
            XCTAssertEqual(decoded, envelope, file.lastPathComponent)
        }
    }

    func testProtocolConstantsMatchManifest() throws {
        let manifest = try readManifest()

        XCTAssertEqual(Set(manifest.commands + manifest.responses), Set(MessageType.all))
    }

    func testInvalidProtocolExamplesAreRejectedByV1Contract() throws {
        let manifest = try readManifest()
        let invalidDirectory = try findInvalidDirectory()
        let invalidFiles = try FileManager.default.contentsOfDirectory(
            at: invalidDirectory,
            includingPropertiesForKeys: nil
        )
            .filter { $0.pathExtension == "json" }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }

        XCTAssertFalse(invalidFiles.isEmpty)

        for file in invalidFiles {
            let root = try XCTUnwrap(JSONSerialization.jsonObject(with: Data(contentsOf: file)) as? [String: Any])
            XCTAssertFalse(isValidEnvelope(root, manifest: manifest), file.lastPathComponent)
        }
    }

    func testFrameFixturesEncodeAndDecodeWithCanonicalBytes() throws {
        let framesDirectory = try findSpecDirectory().appendingPathComponent("frames")
        let frameFiles = try FileManager.default.contentsOfDirectory(
            at: framesDirectory,
            includingPropertiesForKeys: nil
        )
            .filter { $0.pathExtension == "json" }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }

        XCTAssertFalse(frameFiles.isEmpty)

        for file in frameFiles {
            let fixture = try JSONDecoder().decode(FrameFixture.self, from: Data(contentsOf: file))
            let expected = try JSONDecoder().decode(Envelope.self, from: Data(fixture.json.utf8))
            let expectedFrame = try dataFromHex(fixture.frameHex)
            let decoded = try FrameCodec.decodePayload(expectedFrame.dropFirst(4))

            XCTAssertEqual(decoded, expected, file.lastPathComponent)
            XCTAssertEqual(frameLength(expectedFrame), expectedFrame.count - 4, file.lastPathComponent)

            let encoded = try FrameCodec.encode(expected)
            XCTAssertEqual(frameLength(encoded), encoded.count - 4, file.lastPathComponent)
            let redecoded = try FrameCodec.decodePayload(encoded.dropFirst(4))
            XCTAssertEqual(redecoded, expected, file.lastPathComponent)
        }
    }

    private struct ProtocolManifest: Decodable {
        let version: Int
        let commands: [String]
        let responses: [String]
        let errorCodes: [String]
    }

    private struct FrameFixture: Decodable {
        let json: String
        let frameHex: String
    }

    private func readManifest() throws -> ProtocolManifest {
        let manifestURL = try findSpecDirectory().appendingPathComponent("protocol-v1.json")
        return try JSONDecoder().decode(ProtocolManifest.self, from: Data(contentsOf: manifestURL))
    }

    private func isValidEnvelope(_ root: [String: Any], manifest: ProtocolManifest) -> Bool {
        guard let version = root["v"] as? Int,
              let id = root["id"] as? String,
              let type = root["type"] as? String,
              let payload = root["payload"] as? [String: Any],
              version == manifest.version,
              !id.isEmpty,
              !type.isEmpty,
              (manifest.commands + manifest.responses).contains(type) else {
            return false
        }

        switch type {
        case MessageType.hello:
            return stringField(payload, "deviceId", nonEmpty: true) != nil &&
                stringField(payload, "deviceName", nonEmpty: true) != nil
        case MessageType.textInsert:
            return stringField(payload, "text") != nil
        case MessageType.keyTap, MessageType.keyDown, MessageType.keyUp:
            return stringField(payload, "key", nonEmpty: true) != nil
        case MessageType.combo:
            return stringArrayField(payload, "keys")
        case MessageType.mouseMove:
            return intField(payload, "dx") != nil && intField(payload, "dy") != nil
        case MessageType.mouseButton:
            return stringField(payload, "button", nonEmpty: true) != nil &&
                oneOf(payload, "action", values: ["down", "up"])
        case MessageType.mouseClick:
            return stringField(payload, "button", nonEmpty: true) != nil && optionalIntField(payload, "repeat")
        case MessageType.mouseScroll:
            return optionalIntField(payload, "deltaX") && optionalIntField(payload, "deltaY")
        case MessageType.clipboardSet:
            return stringField(payload, "text") != nil
        case MessageType.clipboardGet, MessageType.ping, MessageType.pong:
            return true
        case MessageType.ack:
            return boolField(payload, "ok")
        case MessageType.error:
            guard let code = stringField(payload, "code") else { return false }
            return stringField(payload, "message") != nil && manifest.errorCodes.contains(code)
        case MessageType.authPending:
            return intField(payload, "timeoutSec") != nil && stringField(payload, "message") != nil
        case MessageType.authResult:
            return boolField(payload, "ok") &&
                boolField(payload, "persistToken") &&
                boolField(payload, "trusted") &&
                optionalNullableStringField(payload, "token")
        case MessageType.clipboardValue:
            return stringField(payload, "text") != nil
        case MessageType.shortcutProfile:
            return nullableStringField(payload, "name") && nullableObjectField(payload, "profile")
        default:
            return false
        }
    }

    private func stringField(_ object: [String: Any], _ name: String, nonEmpty: Bool = false) -> String? {
        guard let value = object[name] as? String else { return nil }
        return nonEmpty && value.isEmpty ? nil : value
    }

    private func intField(_ object: [String: Any], _ name: String) -> Int? {
        return object[name] as? Int
    }

    private func optionalIntField(_ object: [String: Any], _ name: String) -> Bool {
        return object[name] == nil || intField(object, name) != nil
    }

    private func boolField(_ object: [String: Any], _ name: String) -> Bool {
        return object[name] as? Bool != nil
    }

    private func stringArrayField(_ object: [String: Any], _ name: String) -> Bool {
        guard let values = object[name] as? [Any] else { return false }
        return values.allSatisfy { $0 is String }
    }

    private func nullableStringField(_ object: [String: Any], _ name: String) -> Bool {
        return object[name] is NSNull || stringField(object, name) != nil
    }

    private func optionalNullableStringField(_ object: [String: Any], _ name: String) -> Bool {
        return object[name] == nil || nullableStringField(object, name)
    }

    private func nullableObjectField(_ object: [String: Any], _ name: String) -> Bool {
        return object[name] is NSNull || object[name] is [String: Any]
    }

    private func oneOf(_ object: [String: Any], _ name: String, values: [String]) -> Bool {
        guard let value = stringField(object, name) else { return false }
        return values.contains(value)
    }

    private func dataFromHex(_ hex: String) throws -> Data {
        let bytes = Array(hex.utf8)
        guard bytes.count.isMultiple(of: 2) else {
            throw NSError(domain: "BlueTypeMacCoreTests", code: 2, userInfo: [NSLocalizedDescriptionKey: "Frame hex has an incomplete byte."])
        }

        var data = Data()
        for index in stride(from: 0, to: bytes.count, by: 2) {
            let pair = String(bytes: bytes[index..<(index + 2)], encoding: .utf8)!
            guard let value = UInt8(pair, radix: 16) else {
                throw NSError(domain: "BlueTypeMacCoreTests", code: 3, userInfo: [NSLocalizedDescriptionKey: "Frame hex contains an invalid byte."])
            }
            data.append(value)
        }
        return data
    }

    private func frameLength(_ frame: Data) -> Int {
        frame.prefix(4).reduce(0) { ($0 << 8) | Int($1) }
    }

    private func findSpecDirectory() throws -> URL {
        var current = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        while true {
            let candidate = current
                .appendingPathComponent("protocol")
                .appendingPathComponent("spec")
            var isDirectory: ObjCBool = false
            if FileManager.default.fileExists(atPath: candidate.path, isDirectory: &isDirectory), isDirectory.boolValue {
                return candidate
            }

            let parent = current.deletingLastPathComponent()
            if parent.path == current.path {
                break
            }
            current = parent
        }

        throw NSError(
            domain: "BlueTypeMacCoreTests",
            code: 1,
            userInfo: [NSLocalizedDescriptionKey: "Could not find protocol/spec from \(FileManager.default.currentDirectoryPath)."]
        )
    }

    private func findExamplesDirectory() throws -> URL {
        return try findSpecDirectory().appendingPathComponent("examples")
    }

    private func findInvalidDirectory() throws -> URL {
        return try findSpecDirectory().appendingPathComponent("invalid")
    }
}
